librz/core/rop.c

```diff
@@ -0,0 +1,1248 @@
+	}
+
+	RzRegInfo *dup = RZ_NEW0(RzRegInfo);
+	if (!dup) {
+		return NULL;
+	}
+
+	dup->name = strdup(src->name);
+	dup->is_mem_read = src->is_mem_read;
+	dup->is_pc_write = src->is_pc_write;
+	dup->is_mem_write = src->is_mem_write;
+	dup->is_var_read = src->is_var_read;
+	dup->is_var_write = src->is_var_write;
+	dup->init_val = src->init_val;
+	dup->new_val = src->new_val;
+
+	return dup;
+}
+
+static bool is_stack_pointer(const RzCore *core, const char *name) {
+	RzRegItem *reg_item = rz_reg_get(core->analysis->reg, name, RZ_REG_TYPE_GPR);
+	if (!reg_item) {
+		return false;
+	}
+	if (core->analysis->bits == 32 && !strcmp(core->analysis->cpu, "x86")) {
+		return !strcmp(reg_item->name, "esp");
+	}
+	if (core->analysis->bits == 64) {
+		return !strcmp(reg_item->name, "rsp") || !strcmp(reg_item->name, "esp");
+	}
+	if (core->analysis->bits == 64 && !strcmp(core->analysis->cpu, "arm")) {
+		return !strcmp(reg_item->name, "r13");
+	}
+	return reg_item->name;
+}
+
+bool is_base_pointer(const RzCore *core, const char *name) {
+	RzRegItem *reg_item = rz_reg_get(core->analysis->reg, name, RZ_REG_TYPE_GPR);
+	if (!reg_item) {
+		return false;
+	}
+	if (core->analysis->bits == 32 && !strcmp(core->analysis->cpu, "x86")) {
+		return !strcmp(reg_item->name, "ebp");
+	}
+	if (core->analysis->bits == 64) {
+		return !strcmp(reg_item->name, "rbp") || !strcmp(reg_item->name, "ebp");
+	}
+	if (core->analysis->bits == 64 && !strcmp(core->analysis->cpu, "arm")) {
+		return !strcmp(reg_item->name, "r11");
+	}
+	return reg_item->name;
+}
+
+void rz_rop_gadget_info_add_dependency(const RzCore *core, RzRopGadgetInfo *gadget_info, const RzILEvent *evt, RzRegInfo *reg_info) {
+	if (!gadget_info || !reg_info) {
+		return;
+	}
+
+	RzRegInfo *reg_info_dup = rz_reg_info_dup(reg_info);
+	if (!reg_info_dup) {
+		return;
+	}
+	switch (evt->type) {
+	case RZ_IL_EVENT_MEM_READ: {
+		const RzILEventMemRead *mem_read = &evt->data.mem_read;
+		reg_info->is_mem_read = true;
+		reg_info->is_mem_write = false;
+		reg_info->is_var_write = false;
+		reg_info_dup->new_val = rz_bv_to_ut64(mem_read->address);
+		break;
+	}
+	case RZ_IL_EVENT_MEM_WRITE: {
+		reg_info->is_mem_write = true;
+		reg_info->is_mem_read = false;
+		reg_info->is_var_write = false;
+		const RzILEventMemWrite *mem_write = &evt->data.mem_write;
+		reg_info_dup->init_val = rz_bv_to_ut64(mem_write->old_value);
+		reg_info_dup->new_val = rz_bv_to_ut64(mem_write->new_value);
+		break;
+	}
+	case RZ_IL_EVENT_VAR_WRITE: {
+		reg_info->is_var_write = true;
+		reg_info->is_mem_read = false;
+		reg_info->is_mem_write = false;
+		const RzILEventVarWrite *var_write = &evt->data.var_write;
+		RzBitVector *init_val = rz_il_value_to_bv(var_write->old_value);
+		RzBitVector *new_val = rz_il_value_to_bv(var_write->new_value);
+		if (!init_val || !new_val) {
+			rz_bv_free(init_val);
+			rz_bv_free(new_val);
+			break;
+		}
+		// reg_info_dup->init_val = rz_bv_to_ut64(init_val);
+		reg_info_dup->new_val = rz_bv_to_ut64(new_val);
```

Comment:
this works, but is also kinda limited to 64 bits. @XVilka what do you think?

Reply:
For now it's ok, but could be improved as more architectures will be added for working with upcoming ROP generator.

Reply:
But please let's do the `instruction word` abstraction first. It is a pain to work with Hexagon and these kind of analysis algos.

Reply:
i think that step should be handled by the disassembler (RzArch) and should be defined by the arch, so we could have a packet with 1 or N instructions.

Reply:
@DMaroo @thestr4ng3r do you have an opinion on this?

Reply:
I will be opening a new issue for tracking this.

Reply:
i think a comment here is enough. just say that this is limited to max 64 bit val but it could be expanded in the future.